core/include/jiminy/core/robot/AbstractConstraint.h

```diff
@@ -0,0 +1,99 @@
+
+        ///////////////////////////////////////////////////////////////////////////////////////////////
+        /// \brief    Compute and return the jacobian of the constraint.
+        ///
+        /// \note     To avoid duplicate kinematic computation, it is assumed that
+        ///           computeJointJacobians and framesForwardKinematics has already
+        ///           been called on model->pncModel_.
+        ///
+        /// \param[in] q    Current joint position.
+        /// \return         Jacobian of the constraint.
+        ///////////////////////////////////////////////////////////////////////////////////////////////
+        virtual matrixN_t const & getJacobian(vectorN_t const & q);
+
+        ///////////////////////////////////////////////////////////////////////////////////////////////
+        /// \brief    Compute and return the drift of the constraint.
+        ///
+        /// \note     To avoid duplicate kinematic computation, it is assumed that forward kinematics
+        ///           and jacobian computation has already been done on model->pncModel_.
+        ///
+        /// \param[in] q    Current joint position.
+        /// \param[in] v    Current joint velocity.
+        /// \return         Drift of the constraint.
+        ///////////////////////////////////////////////////////////////////////////////////////////////
+        virtual vectorN_t const & getDrift(vectorN_t const & q,
+                                           vectorN_t const & v);
+
+    protected:
+        ///////////////////////////////////////////////////////////////////////////////////////////////
+        /// \brief      Link the constraint on the given model, and initialize it.
+        ///
+        /// \param[in] model    Model on which to apply the constraint.
+        /// \return     Error code: attach may fail, including if the constraint is already attached.
+        ///////////////////////////////////////////////////////////////////////////////////////////////
+        virtual hresult_t attach(Model const * model);
+
+        ///////////////////////////////////////////////////////////////////////////////////////////////
+        /// \brief      Detach the constraint from its model.
+        ///////////////////////////////////////////////////////////////////////////////////////////////
+        virtual void detach();
+
+        ///////////////////////////////////////////////////////////////////////////////////////////////
+        /// \brief    Refresh the proxies.
+        ///
+        /// \remark   This method is not intended to be called manually. The Robot to which the
+        ///           motor is added is taking care of it when its own `refresh` method is called.
+        ///////////////////////////////////////////////////////////////////////////////////////////////
+        virtual hresult_t refreshProxies(void);
+
+        Model const * model_; ///< Model on which the constraint operates.
```

Comment:
model_ is already protected.

Reply:
isAttached_ should be private with a const getter (getIsAttached).

Reply:
I don't really see why isAttached_ in particular should be private - I need access to this variable in a child class, a getter and setter for a simple bool seems kind of overkill.

Reply:
I though the derived class only needed a read-only access to this member. It would be the case to force the user to call the base implementation setting the flag, thereby ensuring the desired minimal attaching behaving.

Reply:
Of course no setter should be implemented, otherwise it is irrelevant not to give access to the member directly.

Reply:
No the derived needs access, because initialize or refreshProxies may fail (e.g. if the frame name does not exist in the model) - this is something the parent has no knowledge about.